crates/accelerate/src/barrier_before_final_measurement.rs

```diff
@@ -30,39 +30,87 @@ pub fn barrier_before_final_measurements(
     dag: &mut DAGCircuit,
     label: Option<String>,
 ) -> PyResult<()> {
-    let is_exactly_final = |inst: &PackedInstruction| FINAL_OP_NAMES.contains(&inst.op.name());
-    let final_ops: HashSet<NodeIndex> = dag
-        .op_nodes(true)
-        .filter_map(|(node, inst)| {
-            if !is_exactly_final(inst) {
-                return None;
-            }
-            dag.bfs_successors(node)
-                .all(|(_, child_successors)| {
-                    child_successors.iter().all(|suc| match dag[*suc] {
-                        NodeType::Operation(ref suc_inst) => is_exactly_final(suc_inst),
+    let find_final_nodes = |[_in_index, out_index]: &[NodeIndex; 2]| -> Vec<NodeIndex> {
+        let mut next_nodes: Vec<NodeIndex> = dag
+            .quantum_predecessors(*out_index)
+            .filter(|index| {
+                let node = &dag[*index];
+                match node {
+                    NodeType::Operation(inst) => {
+                        if inst.op.name() == "measure" || inst.op.name() == "barrier" {
+                            dag.bfs_successors(*index).all(|(_, child_successors)| {
+                                child_successors.iter().all(|suc| match &dag[*suc] {
+                                    NodeType::Operation(suc_inst) => {
+                                        suc_inst.op.name() == "measure"
+                                            || suc_inst.op.name() == "barrier"
+                                    }
+                                    _ => true,
+                                })
+                            })
+                        } else {
+                            false
+                        }
+                    }
```

Comment:
Along this method and subsequent ones, we check the same condition (``Node is OpNode && Node.op.name in {"measure", "barrier"}``). Would it be possible to make this another closure? Or would it be inaccessible from within this one?

Reply:
I rewrote the logic to use StandardInstruction matching instead of string equality in: https://github.com/Qiskit/qiskit/pull/13411/commits/c3dbc724c9e063e415f969910ad5fb7ff44430df